### DISCUSSION TITLE
Server Sent Events

### DISCUSSION THREAD

**Discussion by dsyer** [wimdeblauwe/htmx-spring-boot#54] [Ideas]
Maybe I missed it, but I didn't see any support for SSE. HTMX has it built into the client side, and Spring has SSE support on the server side, so it might make sense to try and include it here?

**Comment by wimdeblauwe** [comment]
I don't know what this library should offer about SSE? I have used SSE with Spring and htmx succesfully already.

**Reply by dsyer** [reply]
Isn't it a bit difficult to get HTML rendered in an SSE? I haven't found a nice way to do it in MVC at all. Do you have a sample?

**Reply by wimdeblauwe** [reply]
Ah, yes, you are right. 

There are 2 options I have explorer so far:

1) Send a simple String over SSE and publish that client side as a custom event. Add a `hx-trigger` that would request the HTML via an extra GET request where it is easy to use Thymeleaf to render the HTML.

2) Actually push HTML over the SSE connection. But note that you can't have new lines in the HTML. You can see an example here: https://github.com/wimdeblauwe/blog-example-code/blob/master/htmx-sse/src/main/java/com/wimdeblauwe/examples/htmxsse/PdfGenerationController.java

I create the HTML in a multiline string in Java, not rendering a Thymeleaf template there as I did not really know how to do it. Do notice the use of backslash (`\`) inside the multiline string to have it rendered all on one line.

**Reply by dsyer** [reply]
Yeah, that was kind of where I ended up. It would be nice to be able to render a `View` as an `ServerSentEvent`. I opened this to start looking at options in Spring: https://github.com/spring-projects/spring-framework/issues/30965. There are workarounds, so we can make progress here independently maybe if you are interested.

**Reply by wimdeblauwe** [reply]
I just implemented SSE again in another project and was reminded of this. i am very interested in workaround if you know of any :)

**Reply by dsyer** [reply]
The easiest problem I had to workaround was the one you linked to - there is no support in MVC (although it works in WebFlux) for SSE rendering that spans multiple lines - you have to post process the output and add the data: prefix to every line other than the first. The workaround is shown in the sample project linked from the issue above. There are other issues that need more effort - a `View` needs an `HttpServletRequest/Response` to render but it's hard to dig those out of the context in an `SSEEmitter` for example (my solution to that has been to wrap the `View` in an interceptor with an object that carries the request and response and pull that out when you need it). See also https://github.com/spring-projects/spring-framework/issues/27652 and code snippets in there, and there's a sample linked at the bottom which works last time I looked (checkout the branches for various versions including MVC variants).

**Reply by wimdeblauwe** [reply]
It is weird that you say it works with WebFlux, so you mean using `Flux<ServerSentEvent<String>>` as a return type right? I still had to do a `string.replace("\n", "<br>")` to make it work (pushing HTML over the SSE) with multiline strings.

**Reply by dsyer** [reply]
I don’t think that’s what the sample does. I think `Flux<String>` with an SSE type in the “produces” is what works.

**Reply by wimdeblauwe** [reply]
I had a look to [CompositeViewRenderer.java](https://github.com/dsyer/spring-todo-mvc/blob/main/src/main/java/example/todomvc/web/CompositeViewRenderer.java) (The webflux version). I added it to a test build of the library and tried in my SSE demo project. I replaced returning `Flux<ServerSentEvent<String>>` with `Flux<ServerSentEvent<Rendering>>` using something like this:

```
ServerSentEvent.<Rendering>builder()
    .event("log-event")
    .data(Rendering.view("index :: progress-log-line")
                               .modelAttribute("messages", events)
                               .build())
    .build()
```

to build the SSE event. But unfortunately nothing appears on the SSE stream in the browser. There is no exception on the Java side as well. Should it work normally like that? Or is more code needed (I must say, I don't know enough about Spring internals to understand what this `CompositeViewRendered` is actually doing)?

**Reply by dsyer** [reply]
Stupid question: is your app a webflux server? If it is then you need a method that is annotated as producing SSE with a return type of `Flux<Rendering>` (there’s no need for the `ServerSentEvent` abstraction. The sample you linked to doesn’t have an SSE endpoint, but this one does: https://github.com/dsyer/spring-boot-js-demo/blob/main/htmx/src/main/java/com/example/jsdemo/JsDemoApplication.java.

**Reply by wimdeblauwe** [reply]
The app is a Spring servlet application where I added the `spring-boot-starter-webflux` dependency just for pushing stuff over SSE. My other controller methods are not using `Flux` or `Mono`.

I now tried with adding `@Import({CompositeViewRenderer.class, ThymeleafReactiveViewResolver.class})` to my application but that does not help.

Will it only work in a "webflux only" application?

If I am not able to use the `ServerSentEvent` abstraction, then how is it possible to set the event name?

**Reply by dsyer** [reply]
It needs to be a webflux app to use that specific component. I had one for MVC as well (in a branch) but I’m not sure it works with SSE. I’m sure it could be fixed.

I see what you mean about the event name. I never needed to set it, so it wasn’t a concern, but I see that it should be.

**Reply by wimdeblauwe** [reply]
What I did now to be able to use a Thymeleaf fragment in my Server-sent event is the following:

. Inject an instance of `org.thymeleaf.spring6.SpringTemplateEngine` into my controller
. Use `templateEngine.process("my-template", Set.of("my-fragment"), context).replaceAll("\n", "\ndata:")`

I add the needed variables for the fragment to the Context object.

The `replaceAll` is needed because the Thymeleaf fragment has newlines.

Once you know it, it is not that bad, but if you know of better ways, I am always eager to learn :)

**Reply by checketts** [reply]
That would be worth documenting in the readme

**Reply by wimdeblauwe** [reply]
Not sure if it really belongs in the readme of this library, but it would sure make a good blog post. I'll also add it to my upcoming book on using htmx with Thymeleaf.
